File: src/Flagrum.Core/Mathematics/OrientedBoundingBox.cs
using System.Numerics;
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Mathematics;

public class OrientedBoundingBox
{
    public OrientedBoundingBox() { }

    public OrientedBoundingBox(Vector3 center, Vector3 xHalfExtent, Vector3 yHalfExtent, Vector3 zHalfExtent)
    {
        Center = center;
        XHalfExtent = xHalfExtent;
        YHalfExtent = yHalfExtent;
        ZHalfExtent = zHalfExtent;
    }

    public Vector3 Center { get; set; }
    public Vector3 XHalfExtent { get; set; }
    public Vector3 YHalfExtent { get; set; }
    public Vector3 ZHalfExtent { get; set; }

    public void Read(MessagePackReader reader)
    {
        Center = reader.ReadVector3();
        XHalfExtent = reader.ReadVector3();
        YHalfExtent = reader.ReadVector3();
        ZHalfExtent = reader.ReadVector3();
    }

    public void Write(MessagePackWriter writer)
    {
        Center.Write(writer);
        XHalfExtent.Write(writer);
        YHalfExtent.Write(writer);
        ZHalfExtent.Write(writer);
    }
}

File: src/Flagrum.Blender/core/mathematics/orientedboundingbox.py
from dataclasses import dataclass

from .vector3 import Vector3
from ..serialization.messagepack.reader import MessagePackReader
from ..serialization.messagepack.writer import MessagePackWriter


@dataclass(init=False)
class OrientedBoundingBox:
    center: Vector3
    x_half_extent: Vector3
    y_half_extent: Vector3
    z_half_extent: Vector3

    def read(self, reader: MessagePackReader):
        self.center = reader.read_vector()
        self.x_half_extent = reader.read_vector()
        self.y_half_extent = reader.read_vector()
        self.z_half_extent = reader.read_vector()

    def write(self, writer: MessagePackWriter):
        writer.write_vector(self.center)
        writer.write_vector(self.x_half_extent)
        writer.write_vector(self.y_half_extent)
        writer.write_vector(self.z_half_extent)
